consensus/driver/driver_test.go

```diff
@@ -0,0 +1,223 @@
+		PrevoteBroadcaster:   &mockBroadcaster[tendermint.Prevote[felt.Felt, felt.Felt]]{},
+		PrecommitBroadcaster: &mockBroadcaster[tendermint.Precommit[felt.Felt, felt.Felt]]{},
+	}
+}
+
+func mockTimeoutFn(step tendermint.Step, round tendermint.Round) time.Duration {
+	return 1 * time.Millisecond
+}
+
+func getRandMessageHeader() tendermint.MessageHeader[felt.Felt] {
+	return tendermint.MessageHeader[felt.Felt]{
+		Height: tendermint.Height(rand.Uint32()),
+		Round:  tendermint.Round(rand.Int()),
+		Sender: felt.FromUint64(rand.Uint64()),
+	}
+}
+
+func getRandProposal() tendermint.Proposal[value, felt.Felt, felt.Felt] {
+	return tendermint.Proposal[value, felt.Felt, felt.Felt]{
+		MessageHeader: getRandMessageHeader(),
+		Value:         utils.HeapPtr(value(rand.Uint64())),
+		ValidRound:    tendermint.Round(rand.Int()),
+	}
+}
+
+func getRandPrevote() tendermint.Prevote[felt.Felt, felt.Felt] {
+	return tendermint.Prevote[felt.Felt, felt.Felt]{
+		MessageHeader: getRandMessageHeader(),
+		ID:            utils.HeapPtr(felt.FromUint64(rand.Uint64())),
+	}
+}
+
+func getRandPrecommit() tendermint.Precommit[felt.Felt, felt.Felt] {
+	return tendermint.Precommit[felt.Felt, felt.Felt]{
+		MessageHeader: getRandMessageHeader(),
+		ID:            utils.HeapPtr(felt.FromUint64(rand.Uint64())),
+	}
+}
+
+func getRandTimeout(step tendermint.Step) tendermint.Timeout {
+	return tendermint.Timeout{
+		Height: tendermint.Height(rand.Uint32()),
+		Step:   step,
+		Round:  tendermint.Round(rand.Int()),
+	}
+}
+
+const (
+	actionCount = 10
+)
+
+func registerActions(expectedBroadcast *expectedBroadcast) []tendermint.Action[value, felt.Felt, felt.Felt] {
+	actions := make([]tendermint.Action[value, felt.Felt, felt.Felt], actionCount)
+	for i := range actionCount {
+		switch rand.Int() % 3 {
+		case 0:
+			proposal := getRandProposal()
+			expectedBroadcast.proposals = append(expectedBroadcast.proposals, proposal)
+			actions[i] = utils.HeapPtr(tendermint.BroadcastProposal[value, felt.Felt, felt.Felt](proposal))
+		case 1:
+			prevote := getRandPrevote()
+			expectedBroadcast.prevotes = append(expectedBroadcast.prevotes, prevote)
+			actions[i] = utils.HeapPtr(tendermint.BroadcastPrevote[felt.Felt, felt.Felt](prevote))
+		case 2:
+			precommit := getRandPrecommit()
+			expectedBroadcast.precommits = append(expectedBroadcast.precommits, precommit)
+			actions[i] = utils.HeapPtr(tendermint.BroadcastPrecommit[felt.Felt, felt.Felt](precommit))
+		}
+	}
+	return actions
+}
+
+func toAction(timeout tendermint.Timeout) tendermint.Action[value, felt.Felt, felt.Felt] {
+	return utils.HeapPtr(tendermint.ScheduleTimeout(timeout))
+}
+
+func requireBroadcast[M tendermint.Message[value, felt.Felt, felt.Felt]](
+	expectedBroadcast []M,
+	broadcaster driver.Broadcaster[M, value, felt.Felt, felt.Felt],
+) {
+	broadcaster.(*mockBroadcaster[M]).wg.Add(len(expectedBroadcast))
+}
+
+func assertBroadcast[M tendermint.Message[value, felt.Felt, felt.Felt]](
```

Comment:
nit: feel free to ignore
```suggestion
func waitAndAssertBroadcaster[M tendermint.Message[value, felt.Felt, felt.Felt]](
```